### ISSUE TITLE
Support a unified `oxc.config.ts` for Oxlint and Oxfmt

### ISSUE THREAD

**Issue by usercao** [oxc-project/oxc#25431]
Oxlint and Oxfmt currently use separate configuration files, such as `oxlint.config.ts` and `oxfmt.config.ts`. Since both tools are part of the Oxc toolchain, it would be useful to have an optional, toolchain-level configuration file that can configure both tools in one place.

For example:

```ts
// oxc.config.ts
export default {
  oxfmt: {
    semi: false,
    singleQuote: true,
  },
  oxlint: {
    plugins: ['typescript', 'unicorn', 'oxc'],
    rules: {
      'no-console': 'warn',
    },
  },
}
```

The expected behavior would be:

- `oxfmt` automatically discovers `oxc.config.ts` and reads only the `oxfmt` section.
- `oxlint` automatically discovers `oxc.config.ts` and reads only the `oxlint` section.
- `--config oxc.config.ts` can be used to select the unified configuration explicitly.
- Existing `oxfmt.config.*`, `.oxfmtrc.*`, `oxlint.config.*`, and `.oxlintrc.*` files remain supported for backward compatibility.
- The documentation defines clear precedence when both unified and tool-specific configuration files exist.
- Editor integrations use the same discovery and precedence behavior as the CLIs.

This would provide a single, discoverable entry point for configuring the Oxc toolchain, reduce root-level configuration files, and make it easier to keep related tooling configuration organized in applications, libraries, and monorepos.

Vite+ provides a unified configuration for projects using its toolchain, but native support in Oxlint and Oxfmt would also benefit standalone CLI users and projects that do not use Vite or Vite+.

The exact API and section names are open to discussion. The main request is for Oxlint and Oxfmt to share an optional native configuration entry point while preserving their independent configuration schemas and existing configuration files.

**Comment by connorshea** [comment]
Adding another configuration surface would increase the complexity of the codebase, the number of files we need to look for in the Oxlint and Oxfmt binaries, the VS Code extension, etc, and add more overhead for not much benefit. I don't think it's likely we'll do this, but I'll defer to other team members.

**Comment by usercao** [comment]
Thank you for explaining the concern. I agree that adding another automatically discovered filename across Oxlint, Oxfmt, and the VS Code extension has a real implementation and maintenance cost, and I did not mean to understate that.

I may also have framed the benefit too narrowly as simply reducing the number of files in the repository. The larger benefit, from a user's perspective, is having one ownership boundary for the Oxc toolchain configuration:

- Teams can review linting and formatting policy together in one change.
- Monorepos do not need to maintain two parallel configuration trees at each relevant boundary.
- CLI and editor configuration are less likely to drift apart.
- Shared or organization-wide presets can expose one Oxc entry point while still preserving separate Oxlint and Oxfmt schemas internally.
- Users who install the Oxc VS Code extension for both linting and formatting naturally experience these tools as parts of one toolchain.

This is also why the comparison with the traditional ESLint and Prettier setup feels slightly different to me. ESLint and Prettier are independently developed projects with separate histories, packages, configuration systems, and editor integrations, so separate configuration files are a natural consequence. Oxlint and Oxfmt are still independently usable binaries, but they are presented together as the Oxc tool suite and integrated through a single official Oxc VS Code extension. That gives users a more unified mental model even if the internal implementations remain separate. I do not think this requires the schemas themselves to be merged; only that there could be an optional common container for them.

There is precedent for this UX in comparable toolchains. Biome configures its formatter and linter under the `formatter` and `linter` sections of a single [`biome.json`](https://biomejs.dev/guides/configure-biome/#configuration-file-structure). Vite+ also configures the underlying Oxlint and Oxfmt functionality under `lint` and `fmt` in a single [`vite.config.ts`](https://github.com/voidzero-dev/vite-plus#configuring-vite). I understand that neither is an exact architectural comparison with two independently usable Oxc binaries, but they do suggest that a unified configuration is useful and understandable to users.

Vite+ is a good solution for projects adopting the full Vite+ toolchain, but it is a much larger adoption decision than using Oxlint and Oxfmt directly. A native option would also work for libraries, Node.js projects, non-Vite applications, and users who want the standalone Oxc tools without changing their build and package-management setup.

Would a narrower, explicitly opt-in version address the discovery concern?

For example, the first iteration would not add `oxc.config.ts` to automatic discovery at all. It would only be supported when explicitly passed through the existing `--config` option:

```sh
oxlint --config ./oxc.config.ts .
oxfmt --config ./oxc.config.ts .
```

```ts
// oxc.config.ts
export default {
  oxlint: {
    plugins: ['typescript', 'unicorn', 'oxc'],
    rules: {
      'no-console': 'warn',
    },
  },
  oxfmt: {
    semi: false,
    singleQuote: true,
  },
}
```

Each binary would select only its own section. Existing configuration files and their discovery behavior would remain unchanged. The VS Code extension would not need another default lookup either; users who want the unified file could opt in by pointing the extension at it explicitly. Automatic discovery could remain out of scope unless usage later demonstrates enough demand.

This reduced proposal still introduces a small shared schema/loading surface, so I understand it is not free. However, it avoids increasing the number of files searched by the binaries and editor for every existing project, avoids precedence questions between automatic configuration sources, and provides a way to validate the workflow incrementally.

Another possible approach would be to expose a small shared config-loading API for integrations such as Vite+ and the VS Code extension, while keeping the Oxlint and Oxfmt configuration models independent. I do not have enough context on the internal boundaries to know whether that would actually be simpler, so I am only offering it as an alternative rather than prescribing an implementation.

Thank you for considering this. If even the explicit-only scope is not worth the additional maintenance burden, I understand, but I wanted to clarify the broader use case and suggest a version that directly addresses the configuration-discovery concern.

**Comment by camc314** [comment]
Hi, thanks for the suggestion, closing as not planned.

If you want a unified config file format, please ue vite+. You do not have to use every vite+ feature (test/dev/pkg etc) to take advantage of lint/fmt.

Another option is generated, .gitignored oxlint/oxfmt.config.ts files that you gnenerate from oxc.config.ts files.

Thanks!

**Comment by usercao** [comment]
Thanks for the clarification and for considering the request.

Since this is not planned for Oxc itself, I put together a small temporary workaround for anyone who wants to keep a unified `oxc.config.*` outside of Vite+: [usercao/oxc-config-bridge](https://github.com/usercao/oxc-config-bridge)

It uses one type-safe config source for Oxlint and Oxfmt, forwards to the native CLIs, and can generate editor-compatible proxy configs for the Oxc VS Code extension.

Sharing here in case it is useful to others with a similar workflow.
